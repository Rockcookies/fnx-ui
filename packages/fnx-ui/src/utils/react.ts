import { Children, ReactElement } from 'react';
import { isFragment } from 'react-is';

export const toElementArray = (
	children: React.ReactNode,
	option?: { keepEmpty?: boolean },
): ReactElement[] => {
	const keepEmpty = option && option.keepEmpty;
	let ret: React.ReactElement[] = [];

	Children.forEach(children, (child: any) => {
		if ((child === undefined || child === null) && !keepEmpty) {
			return;
		}

		if (Array.isArray(child)) {
			ret = ret.concat(toElementArray(child));
		} else if (isFragment(child) && child.props) {
			ret = ret.concat(toElementArray(child.props.children, option));
		} else {
			ret.push(child);
		}
	});

	return ret;
};
