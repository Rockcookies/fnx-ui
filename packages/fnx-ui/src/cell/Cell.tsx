import React, {
	createElement,
	forwardRef,
	ForwardRefExoticComponent,
	ReactNode,
	RefAttributes,
} from 'react';
import useProps from '../hooks/use-props';
import { Dictionary } from '../utils/interface';
import { classnames, createBEM } from '../utils/namespace';
import { CellProps } from './interface';

const NS = 'fnx-cell';
const bem = createBEM(NS);

type RequiredCellProps = Required<
	Pick<
		CellProps,
		| 'border'
		| 'itemsAlign'
		| 'clickable'
		| 'titleProps'
		| 'descriptionProps'
		| 'contentProps'
		| 'leftIconProps'
		| 'rightIconProps'
	>
>;

const DEFAULT_PROPS: RequiredCellProps = {
	border: true,
	itemsAlign: 'top',
	clickable: false,
	titleProps: {},
	descriptionProps: {},
	contentProps: {},
	leftIconProps: {},
	rightIconProps: {},
};

const Cell: ForwardRefExoticComponent<
	CellProps & RefAttributes<HTMLDivElement>
> = forwardRef<HTMLDivElement, CellProps>((_props, ref) => {
	const [
		{
			border,
			itemsAlign,
			clickable,
			titleProps,
			descriptionProps,
			contentProps,
			leftIconProps,
			rightIconProps,
		},
		{
			title,
			description,
			content,
			leftIcon,
			rightIcon,
			className,
			children,
			...restProps
		},
	] = useProps<RequiredCellProps, CellProps>(DEFAULT_PROPS, _props);

	const renderNode = (
		tag: string,
		className: string,
		props: Dictionary,
		node: ReactNode,
	): ReactNode => {
		if (node) {
			return createElement(
				tag,
				{ ...props, className: classnames(className, props.className) },
				node,
			);
		}
	};

	return (
		<div
			className={classnames(
				bem([
					`align-${itemsAlign}`,
					{ clickable, 'border-less': !border },
				]),
				className,
			)}
			role={clickable ? 'button' : undefined}
			tabIndex={clickable ? 0 : undefined}
			{...restProps}
			ref={ref}
		>
			{renderNode('span', bem('left-icon'), leftIconProps, leftIcon)}

			{(title || description) && (
				<div
					{...titleProps}
					className={classnames(bem('title'), titleProps.className)}
				>
					{title}
					{renderNode(
						'div',
						bem('description'),
						descriptionProps,
						description,
					)}
				</div>
			)}

			{renderNode(
				'div',
				bem('content', { alone: !(title || description) }),
				contentProps,
				content || children,
			)}

			{renderNode('span', bem('right-icon'), rightIconProps, rightIcon)}
		</div>
	);
});

Cell.displayName = 'Cell';

export default Cell;
