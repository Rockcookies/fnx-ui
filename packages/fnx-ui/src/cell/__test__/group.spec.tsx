import { render } from '@testing-library/react';
import React from 'react';
import Cell from '..';

describe('<Cell.Group />', () => {
	it('should render correctly', () => {
		const { container } = render(
			<Cell.Group title="title">content</Cell.Group>,
		);

		expect(container).toMatchSnapshot();
	});
});
