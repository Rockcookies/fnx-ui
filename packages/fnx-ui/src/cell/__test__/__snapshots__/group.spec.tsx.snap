// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cell.Group /> should render correctly 1`] = `
<div>
  <div
    class="fnx-cell-group"
  >
    <div
      class="fnx-cell-group__title"
    >
      title
    </div>
    <div
      class="fnx-cell-group__body fnx-hairline--top-bottom"
    >
      content
    </div>
  </div>
</div>
`;
